### PR TITLE
Improve nRF Cloud library and sample user experience

### PR DESCRIPTION
**Pull request by plskeggs**
Per IRIS-9151, add detection of which root CA certs are likely to be present, and gate the connection
to the cloud based on what is required for the current connection type (CoAP, MQTT, or REST).

Indicate the FOTA, shadow, and provisioning update intervals more clearly and at INF levels.

Indicate when cloud logging level or alerts are changed by request of the cloud at the INF level.

Slow down provisioning check once the device connects to the cloud.

Jira: IRIS-9151

### CODE REVIEW

File: subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c [hunk 1]
```diff
@@ -493,7 +493,7 @@ int nrf_provisioning_schedule(void)
 	/* To even the load on server side */
 	retry_s += spread_s;
 
-	LOG_DBG("Connecting in %llds", retry_s);
+	LOG_INF("Checking for provisioning commands in %llds seconds", retry_s);
```
Comment (jayteemo) [inline subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c:496]:
```suggestion
	LOG_DBG("Checking for provisioning commands in %llds seconds", retry_s);
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_log.c [hunk 2]
```diff
@@ -319,7 +319,7 @@ void nrf_cloud_log_enable(bool enable)
 		logs_backend_enable(enable);
 #endif
 		enabled = enable;
-		LOG_DBG("enabled = %d", enabled);
+		LOG_INF("Changing cloud logging enabled to:%d", enabled);
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_log.c:322]:
```suggestion
		LOG_DBG("Changing cloud logging enabled to:%d", enabled);
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_log.c [hunk 3]
```diff
@@ -305,7 +305,7 @@ int nrf_cloud_log_control_get(void)
 void nrf_cloud_log_level_set(int level)
 {
 	if (nrf_cloud_log_level != level) {
-		LOG_DBG("Changing log level from:%d to:%d", nrf_cloud_log_level, level);
+		LOG_INF("Changing cloud log level from:%d to:%d", nrf_cloud_log_level, level);
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_log.c:308]:
```suggestion
		LOG_DBG("Changing cloud log level from:%d to:%d", nrf_cloud_log_level, level);
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c [hunk 4]
```diff
@@ -192,27 +241,50 @@ int nrf_cloud_credentials_check(struct nrf_cloud_credentials_status *const cs)
...
 	}
 	cs->prv_key = exists;
 
-	LOG_DBG("Sec Tag: %u, CA: %s, Client Cert: %s, Private Key: %s",
+	LOG_INF("Sec Tag: %u; CA: %s, Client Cert: %s, Private Key: %s",
 		cs->sec_tag,
 		cs->ca		? "Yes" : "No",
 		cs->client_cert	? "Yes" : "No",
 		cs->prv_key	? "Yes" : "No");
+	LOG_INF("CA Size: %zd, Combined: %s, AWS only: %s, CoAP only: %s",
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:283]:
```suggestion
	LOG_DBG("CA Size: %zd, Combined: %s, AWS only: %s, CoAP only: %s",
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c [hunk 5]
```diff
@@ -192,27 +241,50 @@ int nrf_cloud_credentials_check(struct nrf_cloud_credentials_status *const cs)
...
 
 	ret = cred_exists(cs->sec_tag, PRIVATE_KEY, &exists);
 	if (ret < 0) {
+		LOG_ERR("Error checking private key exists");
 		return -EIO;
 	}
 	cs->prv_key = exists;
 
-	LOG_DBG("Sec Tag: %u, CA: %s, Client Cert: %s, Private Key: %s",
+	LOG_INF("Sec Tag: %u; CA: %s, Client Cert: %s, Private Key: %s",
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:278]:
```suggestion
	LOG_DBG("Sec Tag: %u; CA: %s, Client Cert: %s, Private Key: %s",
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:1]: instead of making the logs INF, i would prefer to add a new log module control: `CONFIG_NRF_CLOUD_CREDENTIALS_LOG_LEVEL`
and if you want extra logs in the sample, set the new control to DBG.

Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:1]: same for the nrf_cloud_alert and nrf_cloud_log files

File: samples/cellular/nrf_cloud_multi_service/src/fota_support_coap.c [hunk 6]
```diff
@@ -81,8 +81,8 @@ int coap_fota_thread_fn(void)
 		 */
 		err = nrf_cloud_fota_poll_process(&ctx);
 		if (err == -EAGAIN) {
-			LOG_DBG("Retrying in %d minute(s)",
-				CONFIG_COAP_FOTA_JOB_CHECK_RATE_MINUTES);
+			LOG_INF("Checking for FOTA job in %d seconds",
```
Comment (jayteemo) [inline samples/cellular/nrf_cloud_multi_service/src/fota_support_coap.c:84]: why the change to display in seconds?

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c [hunk 7]
```diff
@@ -177,6 +200,32 @@ static int cred_exists(uint32_t sec_tag, int type, bool *exists)
 	return err;
 }
 
+static int cred_size(uint32_t sec_tag, int type, size_t *cred_sz)
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:203]:
```suggestion
static int cred_size_get(uint32_t sec_tag, int type, size_t *cred_sz)
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c [hunk 8]
```diff
@@ -147,6 +147,29 @@ int nrf_cloud_credentials_configured_check(void)
 		}
 	}
 
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT) || IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		if (!cs.ca_combined && !cs.ca_aws && cs.ca_coap) {
+			/* There is a CA, but not large enough to be correct */
+			LOG_ERR("Cannot connect using MQTT or REST; "
+			        "AWS root CA certificate is missing but CoAP root CA is present.");
+			ret = -ENOTSUP;
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:155]: how about a different return code to indicate that credentials were found, but the size isn't correct for the configured transport type?

File: subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c (hunk 1 above)
Comment (plskeggs) [reply subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c:496]: Read the ticket

File: samples/cellular/nrf_cloud_multi_service/src/fota_support_coap.c (hunk 6 above)
Comment (plskeggs) [reply samples/cellular/nrf_cloud_multi_service/src/fota_support_coap.c:84]: To match the units of measurement used for the shadow logging interval

File: samples/cellular/nrf_cloud_multi_service/src/cloud_connection.c [hunk 9]
```diff
@@ -669,6 +670,13 @@ void cloud_connection_thread_fn(void)
 		if (connect_cloud()) {
 			LOG_DBG("Monitoring nRF Cloud connection");
 
+			if (IS_ENABLED(CONFIG_NRF_PROVISIONING)) {
+				LOG_INF("Reducing provisioning check interval to %d minutes",
+					CONFIG_POST_PROVISIONING_INTERVAL_M);
+				nrf_provisioning_set_interval(CONFIG_POST_PROVISIONING_INTERVAL_M *
+							      SEC_PER_MIN);
+			}
```
Comment (glarsennordic) [inline samples/cellular/nrf_cloud_multi_service/src/cloud_connection.c:681]: I think this should be done inside the `cloud_ready` function rather than directly in the connect loop here

File: subsys/net/lib/nrf_cloud/Kconfig [hunk 10]
```diff
@@ -107,6 +107,26 @@ config NRF_CLOUD_JWT_SOURCE_CUSTOM
 	help
 	  JWTs are created and signed by the nRF Cloud library, not the modem.
 	  The signing key is obtained from the TLS credentials module.
+
+config NRF_CLOUD_COMBINED_CA_CERT_SIZE_THRESHOLD
+	int "Root CA size above which both AWS and CoAP certs are present"
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/Kconfig:112]: `are present` -> `are assumed to be present`

Comment (glarsennordic) [reply subsys/net/lib/nrf_cloud/Kconfig:112]: ditto below

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_log.c (hunk 2 above)
Comment (plskeggs) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_log.c:322]: This is only logged when the user changes the level compared to what the firmware was built for. This will rarely be logged and is not an issue to be INF. It will improve the user experience so they can see that a change they requested actually happened, especially for our samples. Same for alerts.

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c [hunk 11]
```diff
@@ -192,27 +241,50 @@ int nrf_cloud_credentials_check(struct nrf_cloud_credentials_status *const cs)
...
+	LOG_DBG("Sec Tag: %u; CA: %s, Client Cert: %s, Private Key: %s",
 		cs->sec_tag,
 		cs->ca		? "Yes" : "No",
 		cs->client_cert	? "Yes" : "No",
 		cs->prv_key	? "Yes" : "No");
+	LOG_DBG("CA Size: %zd, Combined: %s, AWS only: %s, CoAP only: %s",
+		cs->ca_size,
+		cs->ca_combined ? "Yes" : "No",
+		cs->ca_aws ? "Yes" : "No",
+		cs->ca_coap ? "Yes" : "No");
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:287]: Instead of yes/no, perhaps use `likely/unlikely`?

We wouldn't want a user to trust these indications implicitly and be misled

File: subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c (hunk 1 above)
Comment (glarsennordic) [reply subsys/net/lib/nrf_provisioning/src/nrf_provisioning.c:496]: I prefer this as INF

**Review by glarsennordic — CHANGES_REQUESTED** [review]
request for changes mainly applies to the changes to MSS

**Review by glarsennordic — APPROVED** [review]

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 12]
```diff
@@ -787,6 +788,10 @@ Modem libraries
   * Added the :kconfig:option:`CONFIG_AT_SHELL_UNESCAPE_LF` Kconfig option to enable reception of multiline AT commands.
   * Updated the :c:func:`at_shell` function to replace ``\n`` with ``<CR><LF>`` if :kconfig:option:`CONFIG_AT_SHELL_UNESCAPE_LF` is enabled.
 
+* :ref:`modem_key_mgmt` library:
+
+  * Updated the modem_key_mgmt_read() function to return the actual size buffer required to read the certificate if the size provided is too small.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:793]:
```suggestion
  * Updated the :c:func:`modem_key_mgmt_read()` function to return the actual size buffer required to read the certificate if the size provided is too small.
```

**Review by peknis — CHANGES_REQUESTED** [review]

**Review by juhaylinen — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Review by anhmolt — APPROVED** [review]

File: include/net/nrf_cloud.h [hunk 13]
```diff
@@ -1147,11 +1151,14 @@ int nrf_cloud_credentials_check(struct nrf_cloud_credentials_status *const cs);
 /**
  * @brief Check if the credentials required for connecting to nRF Cloud exist.
  *        The application's configuration is used to determine which credentials
- *        are required.
+ *        are required. Check the size of the root CA certificates installed
+ *        and deny connection if the required root CA certificate(s) are not
```
Comment (jayteemo) [inline include/net/nrf_cloud.h:1155]:
```suggestion
 *        and return an error code if the size of the root CA certificate(s) is not appropriate for the configured transport type.
```

or something similar, since this function doesn't directly deny the connection.

File: include/net/nrf_cloud.h [hunk 14]
```diff
@@ -1147,11 +1151,14 @@ int nrf_cloud_credentials_check(struct nrf_cloud_credentials_status *const cs);
...
+ *        present.
  *
  * @retval 0 Required credentials exist.
  * @retval -EIO Error checking if credentials exists.
  * @retval -ENOTSUP Required credentials do not exist.
+ * @retval -ENOPROTOOPT Incompatible credentials present.
```
Comment (jayteemo) [inline include/net/nrf_cloud.h:1161]:
```suggestion
 * @retval -ENOPROTOOPT Size of root CA is not appropriate for the configured transport type.
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c [hunk 15]
```diff
@@ -192,27 +243,50 @@ int nrf_cloud_credentials_check(struct nrf_cloud_credentials_status *const cs)
...
+	if (exists) {
+		ret = cred_size_get(cs->sec_tag, CA_CERT, &cs->ca_size);
+		if (ret < 0) {
+			LOG_ERR("Error checking CA size");
+			return -EIO;
+		}
+		/* These flags are approximate and only useful for logging to help diagnose
+		 * possible provisioning mistakes.
+		 */
+		cs->ca_combined = (cs->ca_size > CONFIG_NRF_CLOUD_COMBINED_CA_CERT_SIZE_THRESHOLD);
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:259]: can the combined flag be removed?
if both coap and aws flags are set then that indicates it is combined.

```
if (size > combined_min) {
	coap = 1
	aws = 1
} else if (size <= coap_max) {
	coap = 1
	aws = 0
} else if (size > coap_max) {
	coap = 0
	aws = 1
}
```

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c [hunk 16]
```diff
@@ -147,6 +147,29 @@ int nrf_cloud_credentials_configured_check(void)
 		}
 	}
 
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_MQTT) || IS_ENABLED(CONFIG_NRF_CLOUD_REST)) {
+		if (!cs.ca_combined && !cs.ca_aws && cs.ca_coap) {
+			/* There is a CA, but not large enough to be correct */
+			LOG_ERR("Cannot connect using MQTT or REST; "
+				"AWS root CA certificate is missing but CoAP root CA is present.");
```
Comment (jayteemo) [inline subsys/net/lib/nrf_cloud/src/nrf_cloud_credentials.c:154]: similar to the other "likely" log message change, maybe this should be something like:
`LOG_WRN("Connection using MQTT or REST may fail as the size of the CA cert indicates it is a CoAP root CA");`

same for below.

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 17]
```diff
@@ -829,6 +841,8 @@ Libraries for networking
     * To use nRF Cloud's custom MQTT topics instead of the default AWS topics.
     * MQTT and CoAP transports to use a single unified DNS lookup mechanism that supports IPv4 and IPv6, fallback to IPv4, and handling of multiple addresses returned by :c:func:`getaddrinfo`.
     * The log module in the :file:`nrf_cloud_fota_common.c` file from ``NRF_CLOUD`` to ``NRF_CLOUD_FOTA``.
+    * The :c:func:`nrf_cloud_credentials_configured_check` function to retrieve the size of the root CA, and compare it to thresholds to decide whether the CoAP, AWS, or both root CA certs are present.
+      Use this information to log helpful information and decide whether the root CA cert(s) are compatible with the configured connection type.
```
Comment (peknis) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:845]:
```suggestion
      Use this information to log helpful information and decide whether the root CA certificates are compatible with the configured connection type.
```

**Review by peknis — APPROVED** [review]
Approved with a minor nit.

**Review by tony-le-24 — APPROVED** [review]

**Review by jayteemo — APPROVED** [review]

**Review by tejlmand — COMMENTED** [review]
build system changes looks ok

**Review by tejlmand — APPROVED** [review]
